// react-animated-cursor/dist/index.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('react/jsx-runtime'), require('react')) :
    typeof define === 'function' && define.amd ? define(['react/jsx-runtime', 'react'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, global.AnimatedCursor = factory(global.jsxRuntime, global.React));
})(this, (function (jsxRuntime, react) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */


    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    /**
     * useEventListener
     * Hook for handling EventListeners
     * @param {string} eventName - The name of the event to listen for
     * @param {EventListener} handler - The event handler function
     * @param {HTMLElement|Window} element - The element to attach the event listener to
     * @returns {void}
     */
    function useEventListener(eventName, handler, element) {
        if (element === void 0) { element = window; }
        // Create a ref that stores handler
        var savedHandler = react.useRef(null);
        // Update ref.current value if handler changes.
        react.useEffect(function () {
            savedHandler.current = handler;
        }, [handler]);
        react.useEffect(function () {
            // Make sure element supports addEventListener
            var isSupported = element && element.addEventListener;
            if (!isSupported)
                return;
            // Create event listener that calls handler function stored in ref
            var eventListener = function (event) {
                if (savedHandler.current) {
                    savedHandler.current(event);
                }
            };
            // Add event listener
            element.addEventListener(eventName, eventListener);
            // Remove event listener on cleanup
            return function () {
                element.removeEventListener(eventName, eventListener);
            };
        }, [eventName, element]); // Re-run if eventName or element changes
    }

    var IsDevice = (function () {
        if (typeof navigator === 'undefined')
            return;
        var ua = navigator.userAgent;
        return {
            info: ua,
            Android: function () {
                return ua.match(/Android/i);
            },
            BlackBerry: function () {
                return ua.match(/BlackBerry/i);
            },
            IEMobile: function () {
                return ua.match(/IEMobile/i);
            },
            iOS: function () {
                return ua.match(/iPhone|iPad|iPod/i);
            },
            iPad: function () {
                return (ua.match(/Mac/) &&
                    navigator.maxTouchPoints &&
                    navigator.maxTouchPoints > 2);
            },
            OperaMini: function () {
                return ua.match(/Opera Mini/i);
            },
            /**
             * Any Device
             */
            any: function () {
                return (IsDevice.Android() ||
                    IsDevice.BlackBerry() ||
                    IsDevice.iOS() ||
                    IsDevice.iPad() ||
                    IsDevice.OperaMini() ||
                    IsDevice.IEMobile());
            }
        };
    })();

    /**
     * Cursor Core
     * Replaces the native cursor with a custom animated cursor, consisting
     * of an inner and outer dot that scale inversely based on hover or click.
     *
     * @author Stephen Scaff (github.com/stephenscaff)
     *
     * @param {object} obj
     * @param {array}  obj.clickables - array of clickable selectors
     * @param {string} obj.color - rgb color value
     * @param {number} obj.innerScale - inner cursor scale amount
     * @param {number} obj.innerSize - inner cursor size in px
     * @param {object} obj.innerStyle - style object for inner cursor
     * @param {number} obj.outerAlpha - level of alpha transparency for color
     * @param {number} obj.outerScale - outer cursor scale amount
     * @param {number} obj.outerSize - outer cursor size in px
     * @param {object} obj.outerStyle - style object for outer cursor
     * @param {bool}   obj.showSystemCursor - show/hide system cursor1
     * @param {number} obj.trailingSpeed - speed the outer cursor trails at
     */
    function CursorCore(_a) {
        var _b = _a.clickables, clickables = _b === void 0 ? [
            'a',
            'input[type="text"]',
            'input[type="email"]',
            'input[type="number"]',
            'input[type="submit"]',
            'input[type="image"]',
            'label[for]',
            'select',
            'textarea',
            'button',
            '.link'
        ] : _b, _c = _a.color, color = _c === void 0 ? '220, 90, 90' : _c, _d = _a.innerScale, innerScale = _d === void 0 ? 0.6 : _d, _e = _a.innerSize, innerSize = _e === void 0 ? 8 : _e, innerStyle = _a.innerStyle, _f = _a.outerAlpha, outerAlpha = _f === void 0 ? 0.4 : _f, _g = _a.outerScale, outerScale = _g === void 0 ? 6 : _g, _h = _a.outerSize, outerSize = _h === void 0 ? 8 : _h, outerStyle = _a.outerStyle, _j = _a.showSystemCursor, showSystemCursor = _j === void 0 ? false : _j, _k = _a.trailingSpeed, trailingSpeed = _k === void 0 ? 8 : _k;
        var cursorOuterRef = react.useRef(null);
        var cursorInnerRef = react.useRef(null);
        var requestRef = react.useRef(null);
        var previousTimeRef = react.useRef(null);
        var _l = react.useState({
            x: 0,
            y: 0
        }), coords = _l[0], setCoords = _l[1];
        var _m = react.useState(false), isVisible = _m[0], setIsVisible = _m[1];
        var _o = react.useState(false), isActive = _o[0], setIsActive = _o[1];
        var _p = react.useState(false), isActiveClickable = _p[0], setIsActiveClickable = _p[1];
        var endX = react.useRef(0);
        var endY = react.useRef(0);
        /**
         * Primary Mouse move event
         * @param {number} clientX - MouseEvent.clientX
         * @param {number} clientY - MouseEvent.clientY
         */
        var onMouseMove = react.useCallback(function (event) {
            var clientX = event.clientX, clientY = event.clientY;
            setCoords({ x: clientX, y: clientY });
            cursorInnerRef.current.style.top = "".concat(clientY, "px");
            cursorInnerRef.current.style.left = "".concat(clientX, "px");
            endX.current = clientX;
            endY.current = clientY;
        }, []);
        // Outer Cursor Animation Delay
        var animateOuterCursor = react.useCallback(function (time) {
            if (previousTimeRef.current !== undefined) {
                coords.x += (endX.current - coords.x) / trailingSpeed;
                coords.y += (endY.current - coords.y) / trailingSpeed;
                cursorOuterRef.current.style.top = "".concat(coords.y, "px");
                cursorOuterRef.current.style.left = "".concat(coords.x, "px");
            }
            previousTimeRef.current = time;
            requestRef.current = requestAnimationFrame(animateOuterCursor);
        }, [requestRef] // eslint-disable-line
        );
        // Outer cursor RAF setup / cleanup
        react.useEffect(function () {
            requestRef.current = requestAnimationFrame(animateOuterCursor);
            return function () { return cancelAnimationFrame(requestRef.current); };
        }, [animateOuterCursor]);
        /**
         * Calculates amount to scale cursor in px3
         * @param {number} orignalSize - starting size
         * @param {number} scaleAmount - Amount to scale
         * @returns {String} Scale amount in px
         */
        var getScaleAmount = function (orignalSize, scaleAmount) {
            return "".concat(parseInt(String(orignalSize * scaleAmount)), "px");
        };
        // Scales cursor by HxW
        var scaleBySize = react.useCallback(function (cursorRef, orignalSize, scaleAmount) {
            if (cursorRef) {
                cursorRef.style.height = getScaleAmount(orignalSize, scaleAmount);
                cursorRef.style.width = getScaleAmount(orignalSize, scaleAmount);
            }
        }, []);
        // Mouse Events State updates
        var onMouseDown = react.useCallback(function () { return setIsActive(true); }, []);
        var onMouseUp = react.useCallback(function () { return setIsActive(false); }, []);
        var onMouseEnterViewport = react.useCallback(function () { return setIsVisible(true); }, []);
        var onMouseLeaveViewport = react.useCallback(function () { return setIsVisible(false); }, []);
        useEventListener('mousemove', onMouseMove);
        useEventListener('mousedown', onMouseDown);
        useEventListener('mouseup', onMouseUp);
        useEventListener('mouseover', onMouseEnterViewport);
        useEventListener('mouseout', onMouseLeaveViewport);
        // Cursors Hover/Active State
        react.useEffect(function () {
            if (isActive) {
                scaleBySize(cursorInnerRef.current, innerSize, innerScale);
                scaleBySize(cursorOuterRef.current, outerSize, outerScale);
            }
            else {
                scaleBySize(cursorInnerRef.current, innerSize, 1);
                scaleBySize(cursorOuterRef.current, outerSize, 1);
            }
        }, [innerSize, innerScale, outerSize, outerScale, scaleBySize, isActive]);
        // Cursors Click States
        react.useEffect(function () {
            if (isActiveClickable) {
                scaleBySize(cursorInnerRef.current, innerSize, innerScale * 1.2);
                scaleBySize(cursorOuterRef.current, outerSize, outerScale * 1.4);
            }
        }, [
            innerSize,
            innerScale,
            outerSize,
            outerScale,
            scaleBySize,
            isActiveClickable
        ]);
        // Cursor Visibility Statea
        react.useEffect(function () {
            if (isVisible) {
                cursorInnerRef.current.style.opacity = '1';
                cursorOuterRef.current.style.opacity = '1';
            }
            else {
                cursorInnerRef.current.style.opacity = '0';
                cursorOuterRef.current.style.opacity = '0';
            }
        }, [isVisible]);
        // Click event state updates
        react.useEffect(function () {
            var clickableEls = document.querySelectorAll(clickables.join(','));
            clickableEls.forEach(function (el) {
                if (!showSystemCursor)
                    el.style.cursor = 'none';
                el.addEventListener('mouseover', function () {
                    setIsActive(true);
                });
                el.addEventListener('click', function () {
                    setIsActive(true);
                    setIsActiveClickable(false);
                });
                el.addEventListener('mousedown', function () {
                    setIsActiveClickable(true);
                });
                el.addEventListener('mouseup', function () {
                    setIsActive(true);
                });
                el.addEventListener('mouseout', function () {
                    setIsActive(false);
                    setIsActiveClickable(false);
                });
            });
            return function () {
                clickableEls.forEach(function (el) {
                    el.removeEventListener('mouseover', function () {
                        setIsActive(true);
                    });
                    el.removeEventListener('click', function () {
                        setIsActive(true);
                        setIsActiveClickable(false);
                    });
                    el.removeEventListener('mousedown', function () {
                        setIsActiveClickable(true);
                    });
                    el.removeEventListener('mouseup', function () {
                        setIsActive(true);
                    });
                    el.removeEventListener('mouseout', function () {
                        setIsActive(false);
                        setIsActiveClickable(false);
                    });
                });
            };
        }, [isActive, clickables, showSystemCursor]);
        var coreStyles = {
            zIndex: 999,
            display: 'block',
            position: 'fixed',
            borderRadius: '50%',
            pointerEvents: 'none',
            transform: 'translate(-50%, -50%)',
            transition: 'opacity 0.15s ease-in-out, height 0.2s ease-in-out, width 0.2s ease-in-out'
        };
        // Cursor Styles
        var styles = {
            cursorInner: __assign(__assign({ width: innerSize, height: innerSize, backgroundColor: "rgba(".concat(color, ", 1)") }, coreStyles), (innerStyle && innerStyle)),
            cursorOuter: __assign(__assign({ width: outerSize, height: outerSize, backgroundColor: "rgba(".concat(color, ", ").concat(outerAlpha, ")") }, coreStyles), (outerStyle && outerStyle))
        };
        // Hide / Show global cursor
        if (!showSystemCursor)
            document.body.style.cursor = 'none';
        return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx("div", { ref: cursorOuterRef, style: styles.cursorOuter }), jsxRuntime.jsx("div", { ref: cursorInnerRef, style: styles.cursorInner })] }));
    }
    /**
     * AnimatedCursor
     * Calls and passes props to CursorCore if not a touch/mobile device.
     */
    function AnimatedCursor(_a) {
        var clickables = _a.clickables, color = _a.color, innerScale = _a.innerScale, innerSize = _a.innerSize, innerStyle = _a.innerStyle, outerAlpha = _a.outerAlpha, outerScale = _a.outerScale, outerSize = _a.outerSize, outerStyle = _a.outerStyle, showSystemCursor = _a.showSystemCursor, trailingSpeed = _a.trailingSpeed;
        if (typeof navigator !== 'undefined' && IsDevice.any()) {
            return jsxRuntime.jsx(jsxRuntime.Fragment, {});
        }
        return (jsxRuntime.jsx(CursorCore, { clickables: clickables, color: color, innerScale: innerScale, innerSize: innerSize, innerStyle: innerStyle, outerAlpha: outerAlpha, outerScale: outerScale, outerSize: outerSize, outerStyle: outerStyle, showSystemCursor: showSystemCursor, trailingSpeed: trailingSpeed }));
    }

    return AnimatedCursor;

}));
